sdk/core/azure-core/inc/azure/core/http/websockets/websockets.hpp

```diff
@@ -0,0 +1,245 @@
+    WebSocketPeerCloseFrame(uint16_t remoteStatusCode, std::string const& remoteCloseReason)
+        : WebSocketFrame{WebSocketFrameType::PeerClosedReceived},
+          RemoteStatusCode(remoteStatusCode), RemoteCloseReason(remoteCloseReason)
+    {
+    }
+    uint16_t RemoteStatusCode;
+    std::string RemoteCloseReason;
+  };
+
+  struct WebSocketOptions : Azure::Core::_internal::ClientOptions
+  {
+    /**
+     * @brief The set of protocols which are supported by this client
+     */
+    std::vector<std::string> Protocols = {};
+
+    /**
+     * @brief The protocol name of the service client. Used for the User-Agent header
+     * in the initial WebSocket handshake.
+     */
+    std::string ServiceName;
+    /**
+     * @brief The version of the service client. Used for the User-Agent header in the
+     * initial WebSocket handshake
+     */
+    std::string ServiceVersion;
+
+    /**
+     * @brief Construct an instance of a WebSocketOptions type.
+     *
+     * @param protocols Supported protocols for this websocket client.
+     */
+    explicit WebSocketOptions(std::vector<std::string> protocols)
+        : Azure::Core::_internal::ClientOptions{}, Protocols(protocols)
+    {
+    }
+    WebSocketOptions() = default;
+  };
+
+  class WebSocket {
+  public:
+    /** @brief Constructs a new instance of a WebSocket with the specified WebSocket options.
+     *
+     * @param remoteUrl The URL of the remote WebSocket server.
+     * @param options The options to use for the WebSocket.
+     */
+    explicit WebSocket(
+        Azure::Core::Url const& remoteUrl,
+        WebSocketOptions const& options = WebSocketOptions{});
+
+    /** @brief Destroys an instance of a WebSocket.
+     */
+    ~WebSocket();
+
+    /** @brief Opens a WebSocket connection to a remote server.
+     *
+     * @param context Context for the operation, used for cancellation and timeout.
+     */
+    void Open(Azure::Core::Context const& context = Azure::Core::Context{});
+
+    /** @brief Closes a WebSocket connection to the remote server gracefully.
+     *
+     * @param context Context for the operation.
+     */
+    void Close(Azure::Core::Context const& context = Azure::Core::Context{});
+
+    /** @brief Closes a WebSocket connection to the remote server with additional context.
+     *
+     * @param closeStatus 16 bit WebSocket error code.
+     * @param closeReason String describing the reason for closing the socket.
+     * @param context Context for the operation.
+     */
+    void Close(
+        uint16_t closeStatus,
+        std::string const& closeReason = {},
+        Azure::Core::Context const& context = Azure::Core::Context{});
+
+    /** @brief Sends a String frame to the remote server.
+     *
+     * @param textFrame UTF-8 encoded text to send.
+     * @param isFinalFrame if True, this is the final frame in a multi-frame message.
+     * @param context Context for the operation.
+     */
+    void SendFrame(
```

Comment:
Why `SendFrame` is not taking a `ptr<WebSocketFrame>` instead?  Just like `ReceiveFrame` returns a frame, I would expect SendFrame to use a frame

Reply:
The caller of SendFrame knows what they're sending - it's either a string or binary.  So there are overloads for string and binary, which is the only two types of frame which can be sent.
The caller of ReceiveFrame does NOT know what they're getting - it could be binary, could be text, could be a peer closed, etc.

I could add code to create a TextFrame or BinaryFrame, but that means adding a fair amount of complexity to the client code for extremely little benefit.